File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    
    void helper(int strt,int sum,vector<int>&candidates,int target,vector<int>cand,
               vector<vector<int>>&result)
    {
        if(target==sum)
        {
            result.push_back(cand);
            return;
        }
        if(strt==candidates.size())
            return;
        for(int i=strt;i<candidates.size();i++)
        {
            if(sum+candidates[i]>target)
                break;
            if(i!=0&&candidates[i-1]==candidates[i]
               &&(cand.empty()||cand.back()+1!=i))
                continue;
            
            cand.push_back(i);
            helper(i+1,sum+candidates[i],candidates,target,cand,result);
            cand.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<int>cand;
        vector<vector<int>>result;
        helper(0,0,candidates,target,cand,result);
        for(auto &x:result)
        {
            for(auto &y:x )
            {
                y=candidates[y];
            }
        }
        return result;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private void helper(int lvl, int target, int[] candidates, List<Integer>temp, List<List<Integer>> res)
    {
        if(lvl==candidates.length)
        {
            if(target==0)
                res.add(new ArrayList<>(temp));
            return;
        }
        int same=0;
        for(int i=lvl;i<candidates.length;i++)
        {
            if(candidates[lvl]!=candidates[i])break;
            same+=candidates[lvl]==candidates[i]?1:0;
        }
        helper(lvl+same,target,candidates,temp,res);
        for(int i=0;i<same;i++)
        {
            if(target<candidates[lvl]*(i+1))
                break;
            for(int j=0;j<=i;j++)
                temp.add(candidates[lvl]);
            helper(lvl+same,target-candidates[lvl]*(i+1),candidates,temp,res);
            for(int j=0;j<=i;j++)
                temp.remove(temp.size()-1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(0,target,candidates,temp,res);
        return res;
    }
}